### PR TITLE
os/bluestore: log txc details in slow op notification on committed_kv

### PR DESCRIPTION
**Pull request by ifed01**
This might be helpful to troubleshoot issues with slow ops caused by bulky client transactions.

Related-to:  https://tracker.ceph.com/issues/67339

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/kv/RocksDBStore.h [hunk 1]
```diff
@@ -299,6 +299,12 @@ class RocksDBStore : public KeyValueDB {
       const std::string &k,
       const ceph::bufferlist &to_set_bl);
   public:
+    size_t get_count() const {
+      return bat.Count();
+    }
+    size_t get_size_bytes() const {
+      return bat.GetDataSize();
+    }
```
Comment (sajibreadd) [inline src/kv/RocksDBStore.h:307]: `override`?

Comment (ifed01) [reply src/kv/RocksDBStore.h]: yeah, indeed. Thanks!

**Review by sajibreadd — APPROVED** [review]

**Comment by ifed01** [comment]
jenkins test api

**Comment by batrick** [comment]
Adding this to my future runs. Don't wait for me to merge.

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68089](https://tracker.ceph.com/issues/68089).

**Comment by ifed01** [comment]
> @ifed01 I think this is fixing a tracker ticket but it's missing in the commit message. Can you please add it?

@batrick - you mean it's fixed https://tracker.ceph.com/issues/67339?
It couldn't do that - this patch is just extended log output to help to troubleshoot the issue. The latter seems to be flickering...

**Comment by batrick** [comment]
> > @ifed01 I think this is fixing a tracker ticket but it's missing in the commit message. Can you please add it?
> 
> @batrick - you mean it's fixed https://tracker.ceph.com/issues/67339? It couldn't do that - this patch is just extended log output to help to troubleshoot the issue. The latter seems to be flickering...

Ah, my mistake. Maybe add `Test-for: https://tracker.ceph.com/issues/67339`?

**Comment by ifed01** [comment]
> > > @ifed01 I think this is fixing a tracker ticket but it's missing in the commit message. Can you please add it?
> > 
> > 
> > @batrick - you mean it's fixed https://tracker.ceph.com/issues/67339? It couldn't do that - this patch is just extended log output to help to troubleshoot the issue. The latter seems to be flickering...
> 
> Ah, my mistake. Maybe add `Test-for: https://tracker.ceph.com/issues/67339`?

Related-to clause has been added.

**Comment by ifed01** [comment]
jenkins test make check

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68170](https://tracker.ceph.com/issues/68170).

**Comment by batrick** [comment]
> This PR is under test in https://tracker.ceph.com/issues/68170.

@ifed01

https://pulpito.ceph.com/pdonnell-2024-09-20_23:48:22-fs-wip-pdonnell-testing-20240920.212106-debug-distro-default-smithi/7913346/

and a few others to look at with this PR.

**Comment by batrick** [comment]
https://tracker.ceph.com/issues/68170#note-1

**Comment by ifed01** [comment]
> > This PR is under test in https://tracker.ceph.com/issues/68170.
> 
> @ifed01
> 
> https://pulpito.ceph.com/pdonnell-2024-09-20_23:48:22-fs-wip-pdonnell-testing-20240920.212106-debug-distro-default-smithi/7913346/
> 
> and a few others to look at with this PR.

@batrick - nothing bad from a single transaction content perspective for the above run.

I've just added another commit to track maximum amount of pending transactions (and their cost). Having pretty long list of pending transactions could be a cause for slow ops indications as well.
Please try to reproduce the issue with this new patch once again if possible.

**Comment by batrick** [comment]
> > > This PR is under test in https://tracker.ceph.com/issues/68170.
> > 
> > 
> > @ifed01
> > https://pulpito.ceph.com/pdonnell-2024-09-20_23:48:22-fs-wip-pdonnell-testing-20240920.212106-debug-distro-default-smithi/7913346/
> > and a few others to look at with this PR.
> 
> @batrick - nothing bad from a single transaction content perspective for the above run.
> 
> I've just added another commit to track maximum amount of pending transactions (and their cost). Having pretty long list of pending transactions could be a cause for slow ops indications as well. Please try to reproduce the issue with this new patch once again if possible.

I'll keep it in my next batch but not sure when that will be.

@vshankar @rishabh-d-dave @mchangir please include this debugging PR in your future runs.

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68391](https://tracker.ceph.com/issues/68391).

**Comment by batrick** [comment]
@ifed01 note we're ignoring `BLUESTORE_SLOW_OP_ALERT` now via #60011. I will continue including your debugging PR in my batches until you tell me it's no longer necessary.

Anyway, here's some jobs you can look at:

```
 /a/pdonnell-2024-10-04_18:42:30-fs-wip-pdonnell-testing-20241004.144202-debug-distro-default-smithi$ find -name \*ceph.log\* -exec zgrep BLUESTORE_SLOW {} +
./7933201/remote/smithi192/log/9df3a942-8291-11ef-bafb-efdc52797490/ceph.log.gz:2024-10-04T21:03:48.991775+0000 mon.a (mon.0) 1187 : cluster [WRN] Health check failed: 2 OSD(s) experiencing slow operations in BlueStore (BLUESTORE_SLOW_OP_ALERT)
./7933201/remote/smithi192/log/9df3a942-8291-11ef-bafb-efdc52797490/ceph.log.gz:2024-10-04T21:10:00.000638+0000 mon.a (mon.0) 1239 : cluster [WRN] [WRN] BLUESTORE_SLOW_OP_ALERT: 2 OSD(s) experiencing slow operations in BlueStore
./7933201/remote/smithi192/log/9df3a942-8291-11ef-bafb-efdc52797490/ceph.log.gz:2024-10-04T21:12:37.372393+0000 mon.a (mon.0) 1259 : cluster [WRN] Health check update: 3 OSD(s) experiencing slow operations in BlueStore (BLUESTORE_SLOW_OP_ALERT)
./7933201/remote/smithi017/log/9df3a942-8291-11ef-bafb-efdc52797490/ceph.log.gz:2024-10-04T21:03:48.991775+0000 mon.a (mon.0) 1187 : cluster [WRN] Health check failed: 2 OSD(s) experiencing slow operations in BlueStore (BLUESTORE_SLOW_OP_ALERT)
...
```

**Review by aclamk — APPROVED** [review]

**Comment by batrick** [comment]
@ifed01 do you want to merge this into main or keep it only for debugging `wip-` branches?

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68786](https://tracker.ceph.com/issues/68786).

**Pull request merged**